wandb/sdk/launch/agent2/controllers/local_container.py

```diff
@@ -96,36 +98,39 @@ async def reconcile(self):
             # launch it
             await self.launch_item(item)
 
-    async def launch_item(self, item: Any) -> Any:
+        # TODO: clean up orphaned items
+
+    async def launch_item(self, item: Job) -> Optional[str]:
         self.logger.info(f"Launching item: {json.dumps(item, indent=2)}")
 
-        project = LaunchProject.from_spec(item["runSpec"], self.legacy.api)
+        project = LaunchProject.from_spec(item.run_spec, self.legacy.api)
         project.queue_name = self.config["jobset_spec"].name
         project.queue_entity = self.config["jobset_spec"].entity_name
-        project.run_queue_item_id = item["runQueueItemId"]
+        project.run_queue_item_id = item.id
         project.fetch_and_validate_project()
         run_id = project.run_id
         job_tracker = self.legacy.job_tracker_factory(run_id)
         job_tracker.update_run_info(project)
 
-        entrypoint = project.get_single_entry_point()
-        assert entrypoint is not None
-        image_uri = await self.legacy.builder.build_image(
-            project, entrypoint, job_tracker
-        )
-
+        image_uri = None
+        if not project.docker_image:
+            entrypoint = project.get_single_entry_point()
+            assert entrypoint is not None
+            image_uri = await self.legacy.builder.build_image(
+                project, entrypoint, job_tracker
+            )
+        else:
+            assert project.docker_image is not None
+            image_uri = project.docker_image
+
+        assert image_uri is not None
         run = await self.legacy.runner.run(project, image_uri)
         if not run:
             job_tracker.failed_to_start = True
-            self.logger.error(f"Failed to start run for item {item['id']}")
+            self.logger.error(f"Failed to start run for item {item.id}")
             raise NotImplementedError("TODO: handle this case")
 
-        ack_result = await self.queue_driver.ack_run_queue_item(
-            item["runQueueItemId"], run_id
-        )
-        self.logger.info(f"Acked item: {json.dumps(ack_result, indent=2)}")
```

Comment:
Is this meant to be removed or just moved up like in the k8s controller?

Reply:
Oof, I swear that I moved it...